### PR TITLE
[GraphOptz] Add numeric equivalence check to GraphOptz.optimizeBatchN…

### PR DESCRIPTION
**Pull request by mgraczyk**
…ormAfterConvButVarReused

Also simplify test and make it more similar to others in the file.

Summary:
Add numerical equivalence check to GraphOptz.optimizeBatchNormAfterConvButVarReused to ensure that the optimized graph gives the same numerical results as the unoptimized graph.

Test Plan:
Build and run tests a few times to make sure it's not flaky

```
$ for _ in {0..10}; do ninja test; done
```

### CODE REVIEW

File: tests/unittests/GraphOptzTest.cpp [hunk 1]
```diff
@@ -432,40 +440,43 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButConvReused) {
...
 
-  Node *BN = F_->createBatchNormalization("batch", CV, beta, gamma, mean, var,
-                                          3, 0.0001, 0.9);
-  SaveNode *ret = F_->createSave("ret", BN);
-  SaveNode *filterSave = F_->createSave("filterSave", CV->getFilter());
+  ConvolutionNode *CV = F_->createConv(bindings_, "conv", A, 16, 5, 1, 2, 1);
+  Node *BN =
+      F_->createBatchNormalization(bindings_, "batch", CV, 3, 0.0001, 0.9);
+  F_->createSave("ret", BN);
+  F_->createSave("filter", CV->getFilter());
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Thanks for simplifying the test case construction code! Can you include screenshots of the original graph before and after your simplification to make sure they are the same? You can dump a dot graph using `Function::dumpDAG` and render it using `graphviz`.

File: tests/unittests/GraphOptzTest.cpp [hunk 2]
```diff
@@ -432,40 +440,43 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButConvReused) {
...
 
   EXPECT_EQ(F_->getNodes().size(), 4);
+  optimizedF_ = optimizeFunction(F_);
+  ASSERT_EQ(A->getNumUsers(), 2);
+
+  auto *optimizedF_bn =
+      findFunctionNodeByName<SaveNode>(optimizedF_, "ret_save");
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Why is this suffixed with `_bn` if it is a `SaveNode`?

File: tests/unittests/GraphOptzTest.cpp [hunk 3]
```diff
@@ -432,40 +440,43 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButConvReused) {
...
-  EXPECT_TRUE(llvm::isa<BatchNormalizationNode>(ret->getInput()));
+  EXPECT_TRUE(llvm::isa<BatchNormalizationNode>(optimizedF_bn->getInput()));
+
   BatchNormalizationNode *batchNorm =
-      llvm::dyn_cast<BatchNormalizationNode>(ret->getInput());
-  EXPECT_EQ(batchNorm, BN);
-  EXPECT_TRUE(batchNorm && batchNorm->getInput() &&
-              batchNorm->getInput().getNode() == CV);
+      llvm::dyn_cast<BatchNormalizationNode>(optimizedF_bn->getInput());
+  EXPECT_TRUE(batchNorm && batchNorm->getInput());
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]:
```suggestion
  ASSERT_TRUE(batchNorm);
```

File: tests/unittests/GraphOptzTest.cpp [hunk 4]
```diff
@@ -432,40 +440,43 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButConvReused) {
...
+  auto *newCVNode =
+      llvm::dyn_cast<ConvolutionNode>(batchNorm->getInput().getNode());
+  EXPECT_EQ(newCVNode->getInput().getNode(), CV->getInput().getNode());
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]:
```suggestion
  ASSERT_TRUE(newCVNode);
  EXPECT_EQ(newCVNode->getInput().getNode(), CV->getInput().getNode());
```

File: tests/unittests/GraphOptzTest.cpp [hunk 5]
```diff
@@ -84,6 +84,14 @@ static Function *optimizeFunction(Function *F) {
   return G;
 }
 
+template <typename NodeT = Node>
+static const NodeT *findFunctionNodeByName(const Function *f,
+                                           const char *name) {
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Why `const char*`? Why not `const std::string&` or even better yet, ` const llvm::StringRef`?

File: tests/unittests/GraphOptzTest.cpp [hunk 6]
```diff
@@ -84,6 +84,14 @@ static Function *optimizeFunction(Function *F) {
   return G;
 }
 
+template <typename NodeT = Node>
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Please add documentation explaining what this function does. Look at other static helper functions for examples.

File: tests/unittests/GraphOptzTest.cpp [hunk 7]
```diff
@@ -84,6 +84,14 @@ static Function *optimizeFunction(Function *F) {
...
+static const NodeT *findFunctionNodeByName(const Function *f,
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]:
```suggestion
static const NodeT *findFunctionNodeByName(const Function *F,
```

File: tests/unittests/GraphOptzTest.cpp (hunk 2 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: fixed

File: tests/unittests/GraphOptzTest.cpp [hunk 8]
```diff
@@ -432,40 +440,43 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButConvReused) {
...
+  F_->createSave("filter", CV->getFilter());
 
   EXPECT_EQ(F_->getNodes().size(), 4);
+  optimizedF_ = optimizeFunction(F_);
+  ASSERT_EQ(A->getNumUsers(), 2);
+
+  auto *optimizedF_bn =
+      findFunctionNodeByName<SaveNode>(optimizedF_, "ret_save");
+  auto *optimizedF_filterSave =
+      findFunctionNodeByName<SaveNode>(optimizedF_, "filter_save");
```
[the remark this answers is not in the document]
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: Thanks, fixed by using the node's name

File: tests/unittests/GraphOptzTest.cpp (hunk 1 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: I'm not sure that this does preserve structure, but after the change the test is closer to testing what it actually purports to test. Also all of the other tests here setup their convs in this way.

File: tests/unittests/GraphOptzTest.cpp (hunk 4 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: fixed

File: tests/unittests/GraphOptzTest.cpp (hunk 6 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: thanks, added

File: tests/unittests/GraphOptzTest.cpp (hunk 7 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: fixed

File: tests/unittests/GraphOptzTest.cpp (hunk 5 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: fixed

File: tests/unittests/GraphOptzTest.cpp (hunk 1 above)
Comment (SplitInfinity) [reply tests/unittests/GraphOptzTest.cpp]: I'm pretty sure it does. The only difference between the old and the new method is that all of the `Placeholders` for the `Convolution` and `BatchNormalization` nodes are created inside `Function::createXXX` instead of in the scope of the test. 

It's easy to verify - just print out the graph structure using `Function::dumpDAG` before and after your change and examine it using `graphviz`. It is a good idea to include it in the test plan too.

File: tests/unittests/GraphOptzTest.cpp (hunk 3 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: fixed

**Comment by mgraczyk** [comment]
Thanks @SplitInfinity , I've  made your suggested changes.

File: tests/unittests/GraphOptzTest.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply tests/unittests/GraphOptzTest.cpp]: Ah yes that works too

**Comment by mgraczyk** [comment]
Graph structure before and after changes:

### Before
before

### After
after


The name of the save node and some of the "users" count changed, but besides that it is the same structure.

File: tests/unittests/GraphOptzTest.cpp (hunk 1 above)
Comment (mgraczyk) [reply tests/unittests/GraphOptzTest.cpp]: included below

**Comment by SplitInfinity** [comment]
I'm a bit confused. In your "before" screenshot, I see a `SaveNode` for the output of the convolution and the output of the batch normalization. But the code your deleted creates a `SaveNode` to save the filter input of the convolution. Is there a bug in our graph construction code?
